### ISSUE TITLE
Allow passing dict (as opposed to OrderedDict) to nn.Sequential

### ISSUE THREAD

**Issue by hookxs** [pytorch/pytorch#84751]
Feature request: `torch.nn.Sequential` now accepts `collection.OrderedDict` as a way of naming the modules it contains. However, since vanilla `dict` is ordered in python>=3.7 and pytorch currently targets python>=3.7, passing a `dict` should be accepted (ideally, any map str->nn.Module). Currently this throws an error (nn.Sequential insists on collections.OrderedDict).

**Comment by george-qi** [comment]
Thanks for the request! We would accept a PR that fixes this if you'd like to create it

**Comment by albanD** [comment]
I'm not sure we do.
While the two are ordered they do convey different meaning.
In particular, we do have a few Modules that behave differently if they're given a dict() vs an OrderedDict().

**Comment by NicolasHug** [comment]
>  In particular, we do have a few Modules that behave differently if they're given a dict() vs an OrderedDict().

Would you have examples of such modules @albanD ?

**Comment by albanD** [comment]
ParameterDict is one such example: https://github.com/pytorch/pytorch/blob/ef8b91fec73884f3043da8f541176ab7b4c57364/torch/nn/modules/container.py#L824-L829

Not that I advocate for that behavior, just that it is there historically.

**Comment by NicolasHug** [comment]
Thanks for the details @albanD .
It seems clear that in the case of `ParameterDict`, we can't really make `dict` and `OrderedDict` behave the same without a BC break.

However in the case of `Sequential`, `dict` is currently not supported and throws an error. So it seems like we could add support for `dict` (while still keeping `OrderedDict` support) seamlessly without any BC issue?

**Comment by albanD** [comment]
I think the main question I would have here is: what should be the semantic when passing a `dict` to Sequential?
- The same as `OrderedDict` as they are both ordered, this can easily be done.
- The same as done in `ParameterDict` we insert things in the sorted order of the keys.

While the first one makes more sense by itself, lack of consistency here is a problem I think. I would expect that all nn APIs handle these objects in a consistent manner. In particular there are a few places in container.py where we insist on the difference between dict and Ordered dict.

So I am not sure which direction to go into here: should we move globally to dict == OrderedDict and stop doing this distinction? Or do we want to keep globally dict != OrderedDict and we use this to continue explicitly encoding if the order of the entries is important or not?

cc @ezyang in case you have a strong opinion on this.

**Comment by ezyang** [comment]
Can't we just make ParameterDict respect the original dict ordering. It used to not do that but that's because this predates Python 3 making dicts ordered. The dict ordering seems like the right thing.

**Comment by albanD** [comment]
So the move to dict == OrderedDict globally! That's going to be quite a few typing changes and a couple BC-breaking changes.
If someone wants to work on this, I am happy to collaborate in drafting a plan for all of these.

**Comment by jbschlosser** [comment]
I see the following areas within modules that currently interact with dict-likes in public APIs:
* `nn.Sequential`
    * Constructor: doesn't accept `dict` now -> change to accept `dict`, following dict ordering with no sorting (not BC-breaking because this errors now)
* `nn.ModuleDict`
    * Constructor / `update()` method: accepts dicts or OrderedDicts, but treats them the [same](https://github.com/pytorch/pytorch/blob/6650aac8cec0c4d20106c17bb0ce32e59e33e191/torch/nn/modules/container.py#L531-L533) (no BC-breaking change needed)
* `nn.ParameterDict`
    * Constructor / `update()` method: accepts dicts or OrderedDicts, and treats them [differently](https://github.com/pytorch/pytorch/blob/6650aac8cec0c4d20106c17bb0ce32e59e33e191/torch/nn/modules/container.py#L869-L874) (BC-breaking change needed to stop sorting for the case of `dict`)
* `nn.Module`
    * `load_state_dict()`: accepts a `Mapping` and [follows its ordering](https://github.com/pytorch/pytorch/blob/3b7140d938ceebf03e249e1651960b988de9b614/torch/nn/modules/module.py#L2002) (no BC-breaking change needed)

TL;DR: It looks like `nn.ParameterDict` is the only current place that `OrderedDict`s are explicitly treated differently than `dict`s or other `Mapping`s. It's already inconsistent with `nn.ModuleDict`, so I think a BC-breaking change to bring it in line makes sense.

For a *full* global change (i.e. purge `OrderedDict`) from the module code entirely, we'd have to swap types in much more places (e.g. all of the internal module state is currently stored in `OrderedDict`s). Not sure I see the benefit to doing this 93271 

I'll suggest sticking with:
1. Make the BC-breaking change in `nn.ParameterDict` -> treat `Mapping`s the same way as `OrderedDict`s (follow natural ordering)
2. Expand `nn.Sequential` support to handle `Mapping`s in addition to `OrderedDict`s, following natural ordering of the mapping

For both of these, it may be too weird semantically to handle arbitrary `Mapping`s (although for full consistency among modules, we should either do this or make a BC-breaking change to the others to *stop* handling arbitrary `Mapping`s, which is harder). Alternatively, we can explicitly support only `dict` and `OrderedDict`, knowing both of these will be ordered.

**Comment by albanD** [comment]
The proposal sounds good to me.
Do we want to do silent BC-breaking for ParameterDict or we're removing support for dict for one version and then changing the behavior?

**Comment by cpuhrsch** [comment]
@hookxs In general, truly just out of curiosity, why use nn.Sequental at all? Why not loop through a Python list (or dict) of nn.Modules?

I'm mostly worried whether there's some PyTorch composability feature we're lacking when it comes to interoperability with Python lists or dicts (aside from instantiating an nn.Sequential as raised in this issue).

**Comment by hookxs** [comment]
> Why use nn.Sequental at all? Why not loop through a Python list (or dict) of nn.Modules? I'm mostly worried whether there's some PyTorch composability feature we're lacking.

Hi @cpuhrsch, no, I don't think that is the case. In this situation the Sequential was a submodule of another main module and the main module just used it as is, it didn't investigate if it is a collection that it should loop over - a common setup. This main code I didn't want to or couldn't change, I just supplied the submodule.
I think I wrote a quick and project-specific routine which created the submodule from simplified specification (using some defaults like padding etc) and for debugging (and visualization) purposes I used a dict internally so that the modules are named. Then I was annoyed when it didn't work with Sequential because the `if` which rejects a dict but accepts OrderedDict felt completely unnecessary - it would work as is just fine if the `if` simply weren't there:-)

@jbschlosser @albanD The proposal would work well for me. Thank you for taking the time to look at it!

**Comment by cpuhrsch** [comment]
@hookxs - Ah! Thank you for the context. This makes sense. If you have the time I'm wondering whether you'd be comfortable to share a small code example illustrating your issue? It might help us find other issues related to this. Thank you for your support!

**Comment by jbschlosser** [comment]
Discussed internally: we could identify when users **are** relying on order by checking iter(), keys(), etc.

Updated proposal: loud break for the iter case (to ensure users are aware that the order has changed), silent break for other cases.
